Enable FTZ mode, introduce -fftz-math option

linehill opened the pull request:
Optimize ftz marked instructions by enabling ftz mode in the CPU (tested and supported for x86_64). Introduce -fftz-math option to GCC meant for optimizations to assume ftz semantics on floating-point operations. Currently, the option disables optimizations which cause HSA conformance tests to fail.

Also, updated some failing brig related test.

Commit:
brig enables flush-to-zero mode if supported by the target

brig sets -fftz-math flag on, which is used to disable optimizations
which cause fails in HSA conformance tests. Brig enables this option
by default.

common:
* Added two builtin funtion types for (uint64_t) -> void and
  (void) -> uint64_t for calling libhsail-rt routines.

brig-FE:
* Changed: brig1 sets -fftz-math flag on.
* Changed: When FTZ mode is supported then flush_to_zero is not called
  except for f16 operands and for some hsail builtins (min, max,
  fract).

libhsail-rt:
* Added built-ins for setting and restoring FTZ mode.

Commit:
Updated failing tests (did not fix them all)

variables.hsail has one failing test.

pjaaskel commented on gcc/brig/brigfrontend/brig-basic-inst-handler.cc:
```diff
@@ -735,7 +735,7 @@ brig_basic_inst_handler::build_lower_element_broadcast (tree vec_operand)
 
 tree_code
 brig_basic_inst_handler::get_tree_code_for_hsa_opcode
-  (BrigOpcode16_t brig_opcode, BrigType16_t brig_type) const
```
Why remove const?

pjaaskel commented on gcc/brig/brig-lang.c:
```diff
@@ -128,6 +128,11 @@ brig_langhook_init_options_struct (struct gcc_options *opts)
+  /* Prevent some optimizations that break HSAIL's ftz semantics. For
+     example, X * (+/-)1.0 -> (+/-)X simplification is not valid for
+     denormals. */
+  opts->x_flag_ftz_math = 1;
```
Please create a separate pull request (or at least a commit) which has only the -fftz-math core updates as we need to ask for a wider review for it before committing it to gcc trunk. The flag would assume the runtime switches CPU to a FTZ/DAZ mode or has it on by default and affects only compile-time assumptions of values. That is, it would not inject any runtime calls to toggle the flags (that's gccbrig's job).

pjaaskel commented on gcc/brig/brigfrontend/brig-code-entry-handler.cc:
```diff
@@ -1240,6 +1240,22 @@ brig_code_entry_handler::build_operands (const BrigInstBase &brig_inst)
+  if (ftz)
+    {
+      tree_code opcode
+	= brig_basic_inst_handler::get_tree_code_for_hsa_opcode(brig_inst.opcode, brig_inst.type);
```
Looks like a long line. Why do you use the class name specifier here?

pjaaskel replied:
(ok I saw it was converted to a static function)

pjaaskel commented on gcc/brig/brigfrontend/brig-code-entry-handler.cc:
```diff
@@ -1240,6 +1240,22 @@ brig_code_entry_handler::build_operands (const BrigInstBase &brig_inst)
+      tree_code opcode
+	= brig_basic_inst_handler::get_tree_code_for_hsa_opcode(brig_inst.opcode, brig_inst.type);
+      BrigType16_t brig_inner_type = brig_inst.type & BRIG_TYPE_BASE_MASK;
+      if (opcode == CALL_EXPR
```
This needs a comment why are those cases needing the builtin FTZ.

pjaaskel commented on gcc/brig/brigfrontend/brig-code-entry-handler.h:
```diff
@@ -278,6 +278,9 @@ class brig_basic_inst_handler : public brig_code_entry_handler
 
   size_t operator () (const BrigBase *base);
 
+  static tree_code get_tree_code_for_hsa_opcode (BrigOpcode16_t brig_opcode,
```
Could you just move it as a 'protected' method instead?

pjaaskel commented on gcc/brig/brigfrontend/brig-function.cc:
```diff
@@ -584,6 +584,32 @@ brig_function::emit_launcher_and_metadata ()
 	      build_int_cst (uint32_type_node,
 			     m_parent->m_module_group_variables.size()));
 
+  /* Enable FTZ around the kernel launch, save original FTZ setting. */
```
Please expand the comment: what are you doing here exactly and why.

pjaaskel commented on gcc/common.opt:
```diff
@@ -2266,6 +2266,11 @@ fsingle-precision-constant
 
+fftz-math
+Common Report Var(flag_ftz_math) Optimization
+Treats as if denormal floating points are flushed to zero in floating
```
This help text needs updating. 'Makes optimizations assume denormal floating point values must be flushed to zero in floating operations.' or similar. I think key here is that the optimizations do not assume 'ftz' is optional and its semantics do not have to be preserved.

pjaaskel commented on gcc/testsuite/brig.dg/test/gimple/fbarrier.hsail:
```diff
@@ -65,10 +65,10 @@ prog kernel &kernel(kernarg_u64 %input_ptr, kernarg_u64 %output_ptr)
 /* { dg-final { scan-tree-dump "__hsail_waitfbar \\\(0, __context\\\);" "gimple"} } */
 /* { dg-final { scan-tree-dump "__hsail_initfbar \\\(0, __context\\\);" "gimple"} } */
-/* { dg-final { scan-tree-dump "__hsail_initfbar \\\(32, __context\\\);" "gimple"} } */
+/* { dg-final { scan-tree-dump "__hsail_initfbar \\\(__group_local_offset, __context\\\);" "gimple"} } */
```
Sorry! Forgot to update the tests, it seems.

pjaaskel commented on libhsail-rt/Makefile.am:
```diff
@@ -60,7 +60,8 @@ AM_CPPFLAGS = -I$(srcdir)/rt -I$(srcdir)/include/internal
 	-I $(srcdir)/../libgcc \
-	-I $(MULTIBUILDTOP)../../gcc/include $(PTH_CFLAGS)
+	-I $(MULTIBUILDTOP)../../gcc/include $(PTH_CFLAGS) \
+	-I $(srcdir)/../gcc/config/i386
```
Is there a variable for the target's dir to use here instead so it would not be i386 specific (although ftz mode switching currently works only with i386).

pjaaskel commented on libhsail-rt/rt/arithmetic.c:
```diff
@@ -424,6 +361,9 @@ __hsail_fract_f64 (double a)
 uint32_t
 __hsail_class_f32 (float a, uint32_t flags)
 {
+  // FTZ mode is enabled by default but needs to be disabled
```
I think gcc style prefers C style /* */ comments.

pjaaskel commented on libhsail-rt/rt/ftz.c:
```diff
@@ -0,0 +1,176 @@
+    = (!!ftz_flag << INTEL_FTZ_POS) | (!!daz_flag << INTEL_DAZ_POS);
+  return saved_flags;
+#else
+#  error "Unsupported target."
```
Could it be possible to avoid this error and fall back to the awfully slow mode of always calling the builtins in case CPU mode flag toggling is not implemented for the target?

pjaaskel commented on libhsail-rt/rt/ftz.c:
```diff
@@ -0,0 +1,176 @@
+
+  r.f = a;
+
+  // Only need to flush is the exponent bits are all zeros.
```
Check the comment style here.

pjaaskel commented on libhsail-rt/rt/ftz.c:
```diff
@@ -0,0 +1,176 @@
+{
+  if (isnan (a) || isinf (a) || a == 0.0d)
+    return a;
+
```
Extra empty line here.

pjaaskel commented on libhsail-rt/rt/ftz.h:
```diff
@@ -0,0 +1,50 @@
+/* ftz.c -- Builtins for FTZ routines.
```
expand FTZ here

Commit:
Added common -fftz-math flag

With the flag set on the compiler assumes the floating-point
operations must flush received and resulting subnormal floating-point
values to zero.

linehill commented:
Updated the pull request.
